File: overlay/python/hydra-kernel/src/main/python/hydra/dsl/meta/lib/maps.py
"""Phantom-typed term DSL for the hydra.lib.maps library."""

from hydra.typed import TypedTerm
from hydra.dsl.meta.phantoms import lib_primitive as primitive, lib_primitive1 as primitive1, lib_primitive2 as primitive2, lib_primitive3 as primitive3


def alter(f: TypedTerm, key: TypedTerm, mapping: TypedTerm) -> TypedTerm:
    """Alter a value at a key using a function."""
    return primitive3(f, key, mapping)


def bimap(f: TypedTerm, g: TypedTerm, mapping: TypedTerm) -> TypedTerm:
    """Map a function over the keys and values of a map."""
    return primitive3(f, g, mapping)


def delete(key: TypedTerm, mapping: TypedTerm) -> TypedTerm:
    """Remove a key from a map."""
    return primitive2(key, mapping)


def elems(mapping: TypedTerm) -> TypedTerm:
    """Get the values of a map."""
    return primitive1(mapping)


def empty() -> TypedTerm:
    """Create an empty map."""
    return primitive()


def filter(predicate: TypedTerm, mapping: TypedTerm) -> TypedTerm:
    """Filter a map based on values."""
    return primitive2(predicate, mapping)


def filter_with_key(predicate: TypedTerm, mapping: TypedTerm) -> TypedTerm:
    """Filter a map based on key-value pairs."""
    return primitive2(predicate, mapping)


def find_with_default(default: TypedTerm, key: TypedTerm, mapping: TypedTerm) -> TypedTerm:
    """Lookup a value with a default."""
    return primitive3(default, key, mapping)


def from_list(pairs: TypedTerm) -> TypedTerm:
    """Create a map from a list of key-value pairs."""
    return primitive1(pairs)


def insert(key: TypedTerm, value: TypedTerm, mapping: TypedTerm) -> TypedTerm:
    """Insert a key-value pair into a map."""
    return primitive3(key, value, mapping)


def keys(mapping: TypedTerm) -> TypedTerm:
    """Get the keys of a map."""
    return primitive1(mapping)


def lookup(key: TypedTerm, mapping: TypedTerm) -> TypedTerm:
    """Lookup a value in a map."""
    return primitive2(key, mapping)


def map(f: TypedTerm, mapping: TypedTerm) -> TypedTerm:
    """Map a function over a map."""
    return primitive2(f, mapping)


def map_keys(f: TypedTerm, mapping: TypedTerm) -> TypedTerm:
    """Map a function over the keys of a map."""
    return primitive2(f, mapping)


def member(key: TypedTerm, mapping: TypedTerm) -> TypedTerm:
    """Check if a key is present in a map."""
    return primitive2(key, mapping)


def null(mapping: TypedTerm) -> TypedTerm:
    """Check if a map is empty."""
    return primitive1(mapping)


def singleton(key: TypedTerm, value: TypedTerm) -> TypedTerm:
    """Create a map with a single key-value pair."""
    return primitive2(key, value)


def size(mapping: TypedTerm) -> TypedTerm:
    """Get the size of a map."""
    return primitive1(mapping)


def to_list(mapping: TypedTerm) -> TypedTerm:
    """Convert a map to a list of key-value pairs."""
    return primitive1(mapping)


def union(map1: TypedTerm, map2: TypedTerm) -> TypedTerm:
    """Union two maps, with the first taking precedence."""
    return primitive2(map1, map2)

File: overlay/haskell/hydra-kernel/src/main/haskell/Hydra/Dsl/Meta/Lib/Maps.hs
{-# LANGUAGE FlexibleContexts #-}

-- | Phantom-typed term DSL for the hydra.lib.maps library

module Hydra.Dsl.Meta.Lib.Maps where

import Hydra.Typed
import Hydra.Dsl.AsTerm
import Hydra.Dsl.Meta.Phantoms
import qualified Hydra.Dsl.Terms as Terms

import Data.Map
import qualified Hydra.Dsl.Prims as Prims
import qualified Hydra.Lib.Maps as DefMaps


-- | Alter a value at a key using a function.
alter :: TypedTerm (Maybe v -> Maybe v) -> TypedTerm k -> TypedTerm (Map k v) -> TypedTerm (Map k v)
alter = primitive3 (Prims.primName DefMaps.alter)

-- | Map a function over the keys and values of a map.
bimap :: TypedTerm (k1 -> k2) -> TypedTerm (v1 -> v2) -> TypedTerm (Map k1 v1) -> TypedTerm (Map k2 v2)
bimap = primitive3 (Prims.primName DefMaps.bimap)

-- | Remove a key from a map.
delete :: TypedTerm k -> TypedTerm (Map k v) -> TypedTerm (Map k v)
delete = primitive2 (Prims.primName DefMaps.delete)

-- | Get the values of a map.
elems :: TypedTerm (Map k v) -> TypedTerm [v]
elems = primitive1 (Prims.primName DefMaps.elems)

-- | Create an empty map.
empty :: TypedTerm (Map k v)
empty = primitive (Prims.primName DefMaps.empty)

-- | Filter a map based on values.
filter :: TypedTerm (v -> Bool) -> TypedTerm (Map k v) -> TypedTerm (Map k v)
filter = primitive2 (Prims.primName DefMaps.filter)

-- | Filter a map based on key-value pairs.
filterWithKey :: TypedTerm (k -> v -> Bool) -> TypedTerm (Map k v) -> TypedTerm (Map k v)
filterWithKey = primitive2 (Prims.primName DefMaps.filterWithKey)

-- | Lookup a value with a default.
findWithDefault :: AsTerm t v => t -> TypedTerm k -> TypedTerm (Map k v) -> TypedTerm v
findWithDefault def = primitive3 (Prims.primName DefMaps.findWithDefault) (asTerm def)

-- | Create a map from a list of key-value pairs.
fromList :: TypedTerm [(k, v)] -> TypedTerm (Map k v)
fromList = primitive1 (Prims.primName DefMaps.fromList)

-- | Insert a key-value pair into a map.
insert :: TypedTerm k -> TypedTerm v -> TypedTerm (Map k v) -> TypedTerm (Map k v)
insert = primitive3 (Prims.primName DefMaps.insert)

-- | Get the keys of a map.
keys :: TypedTerm (Map k v) -> TypedTerm [k]
keys = primitive1 (Prims.primName DefMaps.keys)

-- | Lookup a value in a map.
lookup :: AsTerm t k => t -> TypedTerm (Map k v) -> TypedTerm (Maybe v)
lookup k = primitive2 (Prims.primName DefMaps.lookup) (asTerm k)

-- | Map a function over a map.
map :: AsTerm f (v1 -> v2) => f -> TypedTerm (Map k v1) -> TypedTerm (Map k v2)
map f = primitive2 (Prims.primName DefMaps.map) (asTerm f)

-- | Map a function over the keys of a map.
mapKeys :: TypedTerm (k1 -> k2) -> TypedTerm (Map k1 v) -> TypedTerm (Map k2 v)
mapKeys = primitive2 (Prims.primName DefMaps.mapKeys)

-- | Check if a key is present in a map.
member :: TypedTerm k -> TypedTerm (Map k v) -> TypedTerm Bool
member = primitive2 (Prims.primName DefMaps.member)

-- | Check if a map is empty.
null :: TypedTerm (Map k v) -> TypedTerm Bool
null = primitive1 (Prims.primName DefMaps.null)

-- | Create a map with a single key-value pair.
singleton :: TypedTerm k -> TypedTerm v -> TypedTerm (Map k v)
singleton = primitive2 (Prims.primName DefMaps.singleton)

-- | Get the size of a map.
size :: TypedTerm (Map k v) -> TypedTerm Int
size = primitive1 (Prims.primName DefMaps.size)

-- | Convert a map to a list of key-value pairs.
toList :: TypedTerm (Map k v) -> TypedTerm [(k, v)]
toList = primitive1 (Prims.primName DefMaps.toList)

-- | Union two maps, with the first taking precedence.
union :: TypedTerm (Map k v) -> TypedTerm (Map k v) -> TypedTerm (Map k v)
union = primitive2 (Prims.primName DefMaps.union)
